### PR TITLE
Segways

### PR DESCRIPTION
**Pull request by LadyFowls**
Add's the ability to build segways for crew use from the protolathe
https://github.com/Aurorastation/Aurora.3/pull/5431

Remake of this PR due to branch issue

### CODE REVIEW

File: html/changelogs/dragosegway.yml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+# Optional: Remove this file after generating master changelog.  Useful for PR changelogs that won't get used again.
+delete-after: True
+
+# Any changes you've made.  See valid prefix list above.
+# INDENT WITH TWO SPACES.  NOT TABS.  SPACES.
+# SCREW THIS UP AND IT WON'T WORK.
+# Also, all entries are changed into a single [] after a master changelog generation. Just remove the brackets when you add new entries.
+# Please surround your changes in  double quotes ("), as certain characters otherwise screws up compiling. The quotes will not show up in the changelog.
+changes: 
+  - imageadd: "Added SegWays to security."
```
Comment (VTCobaltblood) [inline html/changelogs/dragosegway.yml]: Needless CamelCase. Should just be segways.

File: html/changelogs/dragosegway.yml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  - imageadd: "Added Security Motor Pool."
```
Comment (VTCobaltblood) [inline html/changelogs/dragosegway.yml]: "Added a motor pool to security." Why is this all imageadds?

File: code/modules/vehicles/segway.dm [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+/obj/vehicle/segway/verb/siren()
+	set name = "Hail Suspect"
+	set category = "Vehicle"
+	set src in view(0)
+
+	if(usr.incapacitated()) return
+
+	if(!on)
+		usr.visible_message("\The [usr] attempts to use [src]'s siren but it won't work with the engine off.", "You attempt to use \the [src]'s siren with the engine off, it won't work however.", "You hear a thunk.")
```
Comment (VTCobaltblood) [inline code/modules/vehicles/segway.dm]: "\The [usr] attempts to use the siren of their [src], but fails." - don't see why this should state the reason to an observer.
"You need to enable \the engine of your [src] for this to work."

File: code/modules/vehicles/segway.dm [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+		usr << "You start [src]'s engine."
+		playsound(src, 'sound/misc/bike_start.ogg', 100, 1)
+	else
+		if(cell.charge < charge_use)
+			usr << "[src] is out of power."
+		else
+			usr << "[src]'s engine won't start."
+
+/obj/vehicle/segway/verb/stop_engine()
+	set name = "Stop engine"
```
Comment (VTCobaltblood) [inline code/modules/vehicles/segway.dm]: "Stop the engine"

File: code/modules/vehicles/segway.dm [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+	if(!istype(usr, /mob/living/carbon/human))
+		return
+
+	if(!on)
+		usr << "The engine is already stopped."
+		return
+
+	turn_off()
+	if (!on)
+		usr << "You stop [src]'s engine."
```
Comment (VTCobaltblood) [inline code/modules/vehicles/segway.dm]: "You stop the [src]'s engine."

File: code/modules/vehicles/segway.dm [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+	var/space_speed = 8
+	var/segway_icon = "segway"
+
+	var/datum/effect_system/ion_trail/ion
+	var/kickstand = 1
+	var/obj/item/weapon/key/segwaykey/key
+
+/obj/item/weapon/key/segwaykey
+	name = "key"
+	desc = "A small key fob, a ISD symbol is engraved into it."
```
Comment (VTCobaltblood) [inline code/modules/vehicles/segway.dm]: "A small key. It has the symbol of Nanotrasen's Internal Security Department on it."

File: code/modules/vehicles/segway.dm [hunk 7]
```diff
@@ -0,0 +1,221 @@
+/obj/vehicle/segway
+	name = "security segway"
+	desc = "For the slow security or mall cops"
```
Comment (VTCobaltblood) [inline code/modules/vehicles/segway.dm]: Missing a full stop.

File: code/modules/vehicles/segway.dm [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+
+	turn_on()
+	if (on)
+		usr << "You start [src]'s engine."
+		playsound(src, 'sound/misc/bike_start.ogg', 100, 1)
+	else
+		if(cell.charge < charge_use)
+			usr << "[src] is out of power."
+		else
+			usr << "[src]'s engine won't start."
```
Comment (VTCobaltblood) [inline code/modules/vehicles/segway.dm]: "\The [src]'s engine doesn't start."

File: code/modules/vehicles/segway.dm [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+		usr << "The engine is already running."
+		return
+
+	turn_on()
+	if (on)
+		usr << "You start [src]'s engine."
+		playsound(src, 'sound/misc/bike_start.ogg', 100, 1)
+	else
+		if(cell.charge < charge_use)
+			usr << "[src] is out of power."
```
Comment (VTCobaltblood) [inline code/modules/vehicles/segway.dm]: "\The [src] is out of power."

File: code/modules/vehicles/segway.dm [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+
+/obj/vehicle/segway/MouseDrop_T(var/atom/movable/C, mob/user as mob)
+	if(!load(C))
+		user << "<span class='warning'>You were unable to load \the [C] onto \the [src].</span>"
+		return
+
+/obj/vehicle/segway/attack_hand(var/mob/user as mob)
+	if(user == load)
+		unload(load)
+		user << "You unbuckle yourself from \the [src]"
```
Comment (VTCobaltblood) [inline code/modules/vehicles/segway.dm]: Missing a full stop.

File: code/modules/vehicles/segway.dm [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+	else if(user != load && load)
+		user.visible_message ("[user] starts to unbuckle [load] from \the [src]!")
+		if(do_after(user, 8 SECONDS, act_target = src))
+			unload(load)
+			user <<"You unbuckle [load] from \the [src]"
```
Comment (VTCobaltblood) [inline code/modules/vehicles/segway.dm]: Missing a full stop.

File: code/modules/vehicles/segway.dm [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+			load <<"You were unbuckled from \the [src] by [user]"
```
Comment (VTCobaltblood) [inline code/modules/vehicles/segway.dm]: Missing a full stop.

File: code/modules/vehicles/segway.dm [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+/obj/vehicle/segway/attackby(obj/item/weapon/W as obj, mob/user as mob)
+	if(istype(W, /obj/item/weapon/key/segwaykey))
+		if(!key)
+			user.drop_from_inventory(W,src)
+			key = W
+			verbs += /obj/vehicle/train/cargo/engine/verb/remove_key
+		return
+	..()
+
+/obj/vehicle/segway/turn_on()
```
Comment (skull132) [inline code/modules/vehicles/segway.dm]: This proc is doing something horribly wrong. We're adding start_engine, but we should already have it no...? And we're potentially adding start_engine twice?

File: code/modules/vehicles/segway.dm [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+
+		verbs -= /obj/vehicle/segway/verb/stop_engine
+		verbs += /obj/vehicle/segway/verb/start_engine
+
+		if(on)
+			verbs += /obj/vehicle/segway/verb/stop_engine
+		else
+			verbs += /obj/vehicle/segway/verb/start_engine
+
+/obj/vehicle/train/cargo/engine/turn_off()
```
Comment (skull132) [inline code/modules/vehicles/segway.dm]: Same here as above.

File: code/modules/vehicles/segway.dm [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+		verbs += /obj/vehicle/segway/verb/start_engine
+
+
+
+/obj/vehicle/segway/verb/start_engine()
+	set name = "Start engine"
+	set category = "Vehicle"
+	set src in view(0)
+
+	if(!istype(usr, /mob/living/carbon/human))
```
Comment (skull132) [inline code/modules/vehicles/segway.dm]: Should we also check whether the user is on the segway, or can anyone on the same tile diddle this?

File: code/modules/vehicles/segway.dm [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+	if (!on)
+		usr << "You stop [src]'s engine."
+
+
+/obj/vehicle/segway/verb/siren()
+	set name = "Hail Suspect"
+	set category = "Vehicle"
+	set src in view(0)
+
+	if(usr.incapacitated()) return
```
Comment (skull132) [inline code/modules/vehicles/segway.dm]: See above. Also the sanity checks on these should be consistent. Try the `use_check` proc:
```dm
// Checks if user can use this object. Set use_flags to customize what checks are done.
// Returns 0 if they can use it, a value representing why they can't if not.
// Flags are in `code/__defines/misc.dm`
/atom/proc/use_check(mob/user, use_flags = 0, show_messages = FALSE)
```

File: code/modules/vehicles/segway.dm [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+		playsound(src, 'sound/misc/bike_stand_down.ogg', 50, 1)
+		if(pulledby)
+			pulledby.stop_pulling()
+
+	kickstand = !kickstand
+	anchored = (kickstand || on)
+
+/obj/vehicle/segway/load(var/atom/movable/C)
+	var/mob/living/M = C
+	if(!istype(C)) return 0
```
Comment (skull132) [inline code/modules/vehicles/segway.dm]: You'll want `if (!istype(M))`. Further, don't one-line code like this, it hurts readability.

File: code/modules/vehicles/segway.dm [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+			user <<"You unbuckle [load] from \the [src]"
+			load <<"You were unbuckled from \the [src] by [user]"
+
+/obj/vehicle/segway/relaymove(mob/user, direction)
+	if(user != load || !on || user.incapacitated())
+		return
+	return Move(get_step(src, direction))
+
+/obj/vehicle/segway/Move(var/turf/destination)
+	if(kickstand) return
```
Comment (skull132) [inline code/modules/vehicles/segway.dm]: Please don't one-line code.

File: tools/mapmerge2/map_conflict_fixer/Run Map Conflict Fixer.bat [hunk 19]
```diff
@@ -1,4 +1,4 @@
 @echo off
 SET RELATIVEROOT="../../../"
-python map_conflict_fixer.py %1 %RELATIVEROOT%
+py map_conflict_fixer.py %1 %RELATIVEROOT%
```
Comment (skull132) [inline tools/mapmerge2/map_conflict_fixer/Run Map Conflict Fixer.bat]: uh. This should probably be reverted.

**Comment by dragoslava** [comment]
Changes have been fixed

**Comment by dragoslava** [comment]
!review

File: code/modules/vehicles/segway.dm [hunk 20]
```diff
@@ -0,0 +1,199 @@
...
+	var/mob/living/M = C
+	if(!istype(M)) 
+		return 0
+	if(M.buckled || M.restrained() || !Adjacent(M) || !M.Adjacent(src))
+		return 0
+	return ..(M)
+
+/obj/vehicle/segway/MouseDrop_T(var/atom/movable/C, mob/user as mob)
+	if(!load(C))
+		user << "<span class='warning'>You were unable to load \the [C] onto \the [src].</span>"
```
Comment (skull132) [inline code/modules/vehicles/segway.dm]:
```suggestion
		to_chat(user, "<span class='warning'>You were unable to load \the [C] onto \the [src].</span>")
```

File: code/modules/vehicles/segway.dm [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+		return
+
+/obj/vehicle/segway/attack_hand(var/mob/user as mob)
+	if(user == load)
+		unload(load)
+		user << "You unbuckle yourself from \the [src]."
```
Comment (skull132) [inline code/modules/vehicles/segway.dm]:
```suggestion
		to_chat(user, "You unbuckle yourself from \the [src].")
```

File: code/modules/vehicles/segway.dm [hunk 22]
```diff
@@ -0,0 +1,199 @@
...
+	else if(user != load && load)
+		user.visible_message ("[user] starts to unbuckle [load] from \the [src]!")
+		if(do_after(user, 8 SECONDS, act_target = src))
+			unload(load)
+			user <<"You unbuckle [load] from \the [src]."
```
Comment (skull132) [inline code/modules/vehicles/segway.dm]:
```suggestion
			to_chat(user, "You unbuckle [load] from \the [src].")
```

File: code/modules/vehicles/segway.dm [hunk 23]
```diff
@@ -0,0 +1,199 @@
...
+			load <<"You were unbuckled from \the [src] by [user]."
```
Comment (skull132) [inline code/modules/vehicles/segway.dm]:
```suggestion
			to_chat(load, "You were unbuckled from \the [src] by [user].")
```

File: code/modules/vehicles/segway.dm [hunk 24]
```diff
@@ -0,0 +1,199 @@
...
+	if (use_check(usr, show_messages = FALSE))
+		return
+
+	if(!key || (load && load != usr))
+		return
+
+	if(on)
+		turn_off()
+
+	key.loc = usr.loc
```
Comment (skull132) [inline code/modules/vehicles/segway.dm]:
```suggestion
	key.forceMove(usr.loc)
```

File: code/modules/vehicles/segway.dm [hunk 25]
```diff
@@ -0,0 +1,199 @@
...
+		return
+
+	if(kickstand)
+		usr.visible_message("\The [usr] puts up \the [src]'s kickstand.", "You put up \the [src]'s kickstand.", "You hear a thunk.")
+		playsound(src, 'sound/misc/bike_stand_up.ogg', 50, 1)
+	else
+		if(isturf(loc))
+			var/turf/T = loc
+			if (T.is_hole)
+				usr << "<span class='warning'>You don't think kickstands work here.</span>"
```
Comment (skull132) [inline code/modules/vehicles/segway.dm]:
```suggestion
				to_chat(usr, "<span class='warning'>You don't think kickstands work here.</span>")
```

File: code/modules/vehicles/segway.dm [hunk 26]
```diff
@@ -0,0 +1,199 @@
...
+
+/obj/vehicle/segway/bullet_act(var/obj/item/projectile/Proj)
+	if(buckled_mob && prob(protection_percent))
+		buckled_mob.bullet_act(Proj)
+		return
+	..()
+
+
+/obj/vehicle/segway/Destroy()
+	QDEL_NULL(ion)
```
Comment (skull132) [inline code/modules/vehicles/segway.dm]: Missing deleting the key if present.

```suggestion
	QDEL_NULL(ion)

	if (key)
		QDEL_NULL(key)
```

**Comment by dragoslava** [comment]
!review

**Comment by dragoslava** [comment]
!review

Gonna re-bump this again, its been sitting for awhile and all issues were addressed

**Comment by ben10083** [comment]
Is this still active?

**Pull request closed without merging**